File: enchantgui/api/src/main/java/cc/mewcraft/enchantment/gui/api/UiEnchantAdapter.java
package cc.mewcraft.enchantment.gui.api;

import org.jetbrains.annotations.NotNull;

/**
 * Represents an adapter than converts arbitrary enchantment type to {@link UiEnchant}.
 *
 * @param <E> enchantment type
 * @param <T> enchantment target type
 */
public interface UiEnchantAdapter<E, T> {
    /**
     * Initialize this adapter.
     * <p>
     * The implementation should register all available enchantments
     * via {@link UiEnchantProvider#register(UiEnchant)}.
     */
    void initialize();

    @NotNull UiEnchant adaptEnchantment(E enchantment);

    @NotNull UiEnchantTarget adaptEnchantmentTarget(T enchantmentTarget);
}

File: enchantgui/api/src/main/java/cc/mewcraft/enchantment/gui/api/UiEnchantAdapter.kt
package cc.mewcraft.enchantment.gui.api

/**
 * Represents an adapter than converts arbitrary enchantment type to [UiEnchant].
 *
 * All implementations should be put in the package: cc.mewcraft.enchantment.gui.adapter.
 *
 * The following instances will be injected upon instance construction of this class:
 *  * [UiEnchantPlugin]
 *
 * @param <E> enchantment type
 * @param <T> enchantment target type
 */
interface UiEnchantAdapter<E, T> {
    /**
     * Initialize this adapter.
     *
     * The implementation should register all available enchantments via [UiEnchantProvider.register].
     */
    fun initialize()

    /**
     * Checks whether this adapter can be initialized or not.
     *
     * @return true if this adapter can be initialized; false otherwise
     */
    fun canInitialize(): Boolean

    fun adaptEnchantment(rawEnchant: E): UiEnchant
    fun adaptEnchantmentTarget(rawTarget: T): UiEnchantTarget
}
